### PR TITLE
qbd ccc mapping implementation first commit

### PR DESCRIPTION
**Pull request by DhaaraniCIT**

### CODE REVIEW

File: src/app/shared/shared.module.ts [hunk 1]
```diff
@@ -42,10 +43,17 @@ import { ConfigurationLabelComponent } from './components/qbd/configuration/conf
 import { CalendarModule } from 'primeng/calendar';
 import { CascadeSelectModule } from "primeng/cascadeselect";
 import { AppIntegrationLandingComponent } from './components/apps/app-integration-landing/app-integration-landing.component';
+// Import { IntacctStepperComponent } from './components/apps/intacct-stepper/intacct-stepper.component';
+// Import { IntacctLocationEntityComponent } from './components/apps/intacct-location-entity/intacct-location-entity.component';
+// Import { IntacctConnectorComponent } from './components/apps/intacct-connector/intacct-connector.component';
```
Comment (ashwin1111) [inline src/app/shared/shared.module.ts:48]: why are these removed?

File: src/app/core/models/enum/enum.model.ts [hunk 2]
```diff
@@ -246,11 +246,26 @@ export enum QBDOnboardingState {
...
+  }
+
+  export enum FilterOptions {
+    MAPPED = 'Mapped',
+    UNMAPPED = 'Unmapped',
+    EXPORT_BILLS = 'EXPORT_BILLS',
+    REIMBURSABLE = 'REIMBURSABLE',
+    CREDIT_CARD = 'CREDIT_CARD',
+    EXPORT_CREDIT_CARD_PURCHASES = 'EXPORT_CREDIT_CARD_PURCHASES',
+    EXPORT_JOURNALS = 'EXPORT_JOURNALS'
```
[the remark this answers is not in the document]
Comment (DhaaraniCIT) [reply src/app/core/models/enum/enum.model.ts:268]: these are from master, i added only mapping and dashboard

[the remark this answers is not in the document]
Comment (ashwin1111) [reply src/app/core/models/enum/enum.model.ts:268]: let's remove it off if it's not used

File: src/app/integrations/qbd/main/dashboard/dashboard.component.ts [hunk 3]
```diff
@@ -140,6 +145,7 @@ export class DashboardComponent implements OnInit {
   pageChanges(pageNo: number): void {
     this.isLoading = true;
     this.pageNo = pageNo;
+    this.currentPage = Math.ceil(this.pageNo / this.limit)+1;
```
[the remark this answers is not in the document]
Comment (DhaaraniCIT) [reply src/app/integrations/qbd/main/dashboard/dashboard.component.ts:148]: yes,  while we change the page, the component is reloaded, as we have a api call so the page no is 1 always, even the we are in page 2

File: src/app/integrations/qbd/main/mapping/generic-mapping/generic-mapping.component.ts [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+  }
+
+  mappingSeachingFilter(searchValue: string) {
+    if (searchValue.length > 0) {
+      const results: any[] = this.originMapping.results.filter((mapping) =>
+        mapping.source_value.toLowerCase().includes(searchValue)
+      );
+      this.mappings.results = results;
+    } else {
+      this.callGetMappings();
```
[the remark this answers is not in the document]
Comment (DhaaraniCIT) [reply src/app/integrations/qbd/main/mapping/generic-mapping/generic-mapping.component.ts:62]: if the search box is empty or cleared

File: src/app/shared/shared.module.ts (hunk 1 above)
Comment (DhaaraniCIT) [reply src/app/shared/shared.module.ts:48]: i merged the master to local, so these were imported and throw error, so commented them

File: src/app/shared/components/qbd/configuration/export-setting/export-setting.component.ts [hunk 5]
```diff
@@ -347,6 +349,7 @@ export class ExportSettingComponent implements OnInit {
     this.exportSettingService.postQbdExportSettings(exportSettingPayload).subscribe((response: QBDExportSettingGet) => {
       this.saveInProgress = false;
       this.toastService.displayToastMessage(ToastSeverity.SUCCESS, 'Export settings saved successfully');
+      this.mappingService.refreshMappingPages();
```
[the remark this answers is not in the document]
Comment (DhaaraniCIT) [reply src/app/shared/components/qbd/configuration/export-setting/export-setting.component.ts:352]: for this only we are checking the export setting has ccc export on or off to show the mapping, anyways we will show the dashboard only first.

this service is to monitor if the ccc is turned off and saved, to hide the mapping, and vise versa

[the remark this answers is not in the document]
Comment (ashwin1111) [reply src/app/shared/components/qbd/configuration/export-setting/export-setting.component.ts:352]: if someone doesn't visit this page and hit save how will the mapping tab be visible?

File: src/app/shared/shared.module.ts (hunk 1 above)
Comment (ashwin1111) [reply src/app/shared/shared.module.ts:48]: what error? you cannot remove these

File: src/app/integrations/qbd/main/mapping/generic-mapping/generic-mapping.component.ts [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    this.callGetMappings();
+  }
+
+  pageChanges(pageNo: number): void {
+    this.isLoading = true;
+    this.pageNo = pageNo;
+    this.currentPage = Math.ceil(this.pageNo / this.limit)+1;
+    this.selectedMappingFilter = this.selectedMappingFilter ? this.selectedMappingFilter : null;
+    this.callGetMappings();
+  }
```
[the remark this answers is not in the document]
Comment (ashwin1111) [reply src/app/integrations/qbd/main/mapping/generic-mapping/generic-mapping.component.ts:90]: > the offset will not change when we change the page no

this is not true. Consider you're on page 1 and page size is 10
offset would be 0 first and limit would be 10

When you move to page 2, offset becomes 10, limit 10
on page 3, offset 20, limit 10

File: src/app/shared/components/qbd/mapping/mapping-table/mapping-table.component.html [hunk 7]
```diff
@@ -3,17 +3,17 @@
         <ng-template pTemplate="header">
             <tr>
                 <th>Corporate Card in Fyle</th>
-                <th>QuickBooks Desktop Credit Card Account</th>
+                <th>{{ appName }} Credit Card Account</th>
```
Comment (ashwin1111) [inline src/app/shared/components/qbd/mapping/mapping-table/mapping-table.component.html:6]: the whole name should come as input, we'll use it in other apps and other mapping pages

File: src/app/shared/components/qbd/mapping/mapping-table/mapping-table.component.ts [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    const mappingPayload = MappingModel.constructPayload(data);
+    this.postMapping.emit(mappingPayload);
+    this.isSelectedRow = 0;
+  }
+
+  getToolTip() {
+    if (this.operatingSystem === OperatingSystem.MAC) {
+      return '<div style="padding:0px 6px 4px;text-align: center;><p style="font-size:12px;padding-top:0">Save</p><p style="margin-top:5px;padding:4px;background: #A9ACBC80;font-size:10px;border-radius: 4px;">return<p></div>';
+    }
+    return '<div style="padding:0px 6px 4px;text-align: center;><p style="font-size:12px;padding-top:0">Save</p><p style="margin-top:5px;padding:4px;background: #A9ACBC80;font-size:10px;border-radius: 4px;">⏎<p></div>';
```
Comment (ashwin1111) [inline src/app/shared/components/qbd/mapping/mapping-table/mapping-table.component.ts:59]: why are we not adding the next directly in tooltip text?

File: src/app/shared/components/qbd/mapping/mapping-table/mapping-table.component.ts [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+  isTypingInBox(event: any, row: Mapping) {
+    if (event.keyCode === 13) {
+      this.isSelectedRow = 0;
+      this.saveRow(row);
+    }
+    this.isSelectedRow = row.id;
+  }
+
+  saveRow(data: Mapping) {
```
Comment (ashwin1111) [inline src/app/shared/components/qbd/mapping/mapping-table/mapping-table.component.ts:48]: postTextFieldValue()

File: src/app/shared/components/qbd/mapping/mapping-table/mapping-table.component.ts [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+  @Input() operatingSystem: string;
+
+  @Output() postMapping = new EventEmitter<MappingPost>();
+
+  @ViewChild('customTooltip') customTooltip: any;
+
+  allOS = OperatingSystem;
+
+  isSelectedRow: number;
```
Comment (ashwin1111) [inline src/app/shared/components/qbd/mapping/mapping-table/mapping-table.component.ts:27]: `focussedMappingId`

File: src/app/shared/components/qbd/mapping/mapping-table/mapping-table.component.ts [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+  destinationValue: string;
+
+  fieldType = FieldType;
+
+  constructor() { }
+
+  onTextBoxChange(event: any, index: number) {
+    const tragetValue: string = event.target.value;
```
Comment (ashwin1111) [inline src/app/shared/components/qbd/mapping/mapping-table/mapping-table.component.ts:36]: targetValue

File: src/app/shared/components/qbd/mapping/mapping-table/mapping-table.component.ts [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    this.destinationValue = tragetValue;
```
Comment (ashwin1111) [inline src/app/shared/components/qbd/mapping/mapping-table/mapping-table.component.ts:37]: use form instead of doing this, can use form value in 49th line

File: src/app/shared/components/qbd/mapping/mapping-table/mapping-table.component.ts (hunk 8 above)
Comment (DhaaraniCIT) [reply src/app/shared/components/qbd/mapping/mapping-table/mapping-table.component.ts:59]: we need to add some style for texts that why, and we need to change the text according to OS, that's why

File: src/app/integrations/qbd/main/mapping/generic-mapping/generic-mapping.component.ts [hunk 13]
```diff
@@ -21,6 +20,8 @@ export class GenericMappingComponent implements OnInit {
 
   mappings: MappingResponse;
 
+  filterMappings: Mapping[];
```
Comment (ashwin1111) [inline src/app/integrations/qbd/main/mapping/generic-mapping/generic-mapping.component.ts:23]: `filteredMappings`

File: src/app/shared/components/qbd/mapping/mapping-table/mapping-table.component.html [hunk 14]
```diff
@@ -2,8 +2,8 @@
     <p-table [value]="mappings" [tableStyle]="{ 'min-width': '50rem' }">
         <ng-template pTemplate="header">
             <tr>
-                <th>Corporate Card in Fyle</th>
-                <th>{{ appName }} Credit Card Account</th>
+                <th>{{ headerText1 }}</th>
```
Comment (ashwin1111) [inline src/app/shared/components/qbd/mapping/mapping-table/mapping-table.component.html:5]: `fyleHeaderName`

File: src/app/shared/components/qbd/mapping/mapping-table/mapping-table.component.html [hunk 15]
```diff
@@ -2,8 +2,8 @@
...
+                <th>{{ headerText2 }}</th>
```
Comment (ashwin1111) [inline src/app/shared/components/qbd/mapping/mapping-table/mapping-table.component.html:6]: destinationHeaderName

File: src/app/shared/components/qbd/mapping/mapping-table/mapping-table.component.ts [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+
+  postTextFieldValue(data: Mapping) {
+    data.destination_value = this.destinationValue;
+    const mappingPayload = MappingModel.constructPayload(data);
+    this.postMapping.emit(mappingPayload);
+    this.focussedMappingId = 0;
+  }
+
+  getToolTip() {
+    if (this.operatingSystem === OperatingSystem.MAC) {
```
Comment (ashwin1111) [inline src/app/shared/components/qbd/mapping/mapping-table/mapping-table.component.ts:56]: https://fylein.slack.com/archives/C05J5DBE1GF/p1694508258408509?thread_ts=1694155201.347349&cid=C05J5DBE1GF

File: src/app/shared/components/qbd/mapping/mapping-table/mapping-table.component.ts [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+  }
+
+  postTextFieldValue(data: Mapping) {
+    data.destination_value = this.destinationValue;
+    const mappingPayload = MappingModel.constructPayload(data);
+    this.postMapping.emit(mappingPayload);
+    this.focussedMappingId = 0;
+  }
+
+  getToolTip() {
```
Comment (ashwin1111) [inline src/app/shared/components/qbd/mapping/mapping-table/mapping-table.component.ts:55]: getToolTipText()

File: src/app/shared/components/qbd/mapping/mapping-table/mapping-table.component.ts [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+  getToolTipText() {
+    return `
+            <div style="padding:0px 6px 4px;text-align: center;>
+              <p style="font-size:12px;padding-top:0">Save</p>
+              <p style="margin-top:5px;padding:4px;background: #A9ACBC80;font-size:10px;border-radius: 4px;">${this.operatingSystem === OperatingSystem.MAC ? `return` : `⏎`}<p>
+            </div>`;
+  }
+
+  ngOnInit(): void {
+    this.getToolTipText();
```
Comment (ashwin1111) [inline src/app/shared/components/qbd/mapping/mapping-table/mapping-table.component.ts:64]: this is not required

**Review by ashwin1111 — APPROVED** [review]
